### PR TITLE
Implement delegate_names to allow decorating delegated attributes

### PR DESCRIPTION
**Pull request by jbrockmendel**
This PR defines a `delegate_names` decorator that provides an alternative (and to my taste, much nicer) syntax for pinning delegated attributes to Accessor classes.

Effectively this just moves the call to `_add_delegate_accessors` from after the class definition to a decorator on the class.  I find this much harder to overlook when reading a class definition.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by TomAugspurger** [comment]
If you fetch and merge master the CI failures should be fixed.

**Review by TomAugspurger — APPROVED** [review]
Agreed this is nicer.

Merging on Monday if no objections.

File: pandas/core/indexes/category.py [hunk 1]
```diff
@@ -835,24 +846,8 @@ def _delegate_method(self, name, *args, **kwargs):
...
-                                             "as_ordered", "as_unordered",
-                                             "min", "max"],
-            typ='method', overwrite=True)
-
 
 CategoricalIndex._add_numeric_methods_add_sub_disabled()
 CategoricalIndex._add_numeric_methods_disabled()
 CategoricalIndex._add_logical_methods_disabled()
 CategoricalIndex._add_comparison_methods()
-CategoricalIndex._add_accessors()
```
Comment (jreback) [inline pandas/core/indexes/category.py]: slightly -0 here as I like the ``_add_*`` flow. but ok

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**